File: app/src/main/kotlin/com/myradio/StationRepository.kt
package com.myradio

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import java.net.HttpURLConnection
import java.net.URL
import java.util.Locale

class StationRepository(private val context: Context) {

    suspend fun fetchNearbyStations(existing: List<RadioStation>): List<RadioStation> =
        withContext(Dispatchers.IO) {
            try {
                val countryCode = Locale.getDefault().country.ifEmpty { "BG" }
                Log.d(TAG, "Country code: $countryCode")

                val urlString = "https://all.api.radio-browser.info/json/stations/search" +
                    "?countrycode=$countryCode&hidebroken=true&order=votes&reverse=true&limit=30"

                Log.d(TAG, "Fetching: $urlString")
                val json = fetchJson(urlString) ?: return@withContext emptyList<RadioStation>().also {
                    Log.w(TAG, "fetchJson returned null")
                }
                val result = parseStations(json, existing)
                Log.d(TAG, "Parsed ${result.size} stations from ${json.length()} results")
                result
            } catch (e: Exception) {
                Log.e(TAG, "fetchNearbyStations failed", e)
                emptyList()
            }
        }

    private fun fetchJson(urlString: String): JSONArray? {
        val conn = URL(urlString).openConnection() as HttpURLConnection
        return try {
            conn.setRequestProperty("User-Agent", "MyRadio/1.0")
            conn.connectTimeout = 10_000
            conn.readTimeout = 10_000
            val code = conn.responseCode
            Log.d(TAG, "HTTP $code for $urlString")
            if (code != 200) return null
            val body = conn.inputStream.bufferedReader().readText()
            JSONArray(body)
        } finally {
            conn.disconnect()
        }
    }

    companion object {
        private const val TAG = "StationRepository"
    }

    private fun parseStations(json: JSONArray, existing: List<RadioStation>): List<RadioStation> {
        val existingNames = existing.map { it.name.lowercase() }.toHashSet()
        val existingUrls = existing.map { it.streamUrl }.toHashSet()
        val result = mutableListOf<RadioStation>()
        var nextId = 1000

        for (i in 0 until json.length()) {
            val obj = json.optJSONObject(i) ?: continue
            val name = obj.optString("name").trim()
            if (name.isEmpty()) continue
            val url = obj.optString("url_resolved").trim()
            if (url.isEmpty()) continue
            val country = obj.optString("country", "")

            if (name.lowercase() in existingNames) continue
            if (url in existingUrls) continue

            existingNames.add(name.lowercase())
            existingUrls.add(url)

            result.add(
                RadioStation(
                    id = nextId++,
                    name = name,
                    streamUrl = url,
                    description = country.ifEmpty { obj.optString("tags", "") }
                )
            )
        }
        return result
    }
}

File: MyRadioiOS/StationRepository.swift
import Foundation

/// Fetches additional radio stations from the Radio Browser API.
/// Equivalent to StationRepository on Android.
actor StationRepository {

    static let shared = StationRepository()
    private init() {}

    func fetchNearbyStations(existing: [RadioStation]) async -> [RadioStation] {
        let countryCode = Locale.current.region?.identifier ?? "BG"
        let urlString = "https://all.api.radio-browser.info/json/stations/search"
            + "?countrycode=\(countryCode)&hidebroken=true&order=votes&reverse=true&limit=30"

        guard let url = URL(string: urlString) else { return [] }

        do {
            var request = URLRequest(url: url, timeoutInterval: 10)
            request.setValue("MyRadio/1.0", forHTTPHeaderField: "User-Agent")

            let (data, response) = try await URLSession.shared.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return [] }

            let json = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] ?? []
            return parseStations(json, existing: existing)
        } catch {
            print("StationRepository error: \(error)")
            return []
        }
    }

    private func parseStations(_ json: [[String: Any]], existing: [RadioStation]) -> [RadioStation] {
        var existingNames = Set(existing.map { $0.name.lowercased() })
        var existingUrls  = Set(existing.map { $0.streamUrl })
        var result: [RadioStation] = []
        var nextId = 1000

        for obj in json {
            let name = (obj["name"] as? String ?? "").trimmingCharacters(in: .whitespaces)
            let url  = (obj["url_resolved"] as? String ?? "").trimmingCharacters(in: .whitespaces)
            guard !name.isEmpty, !url.isEmpty else { continue }
            guard !existingNames.contains(name.lowercased()) else { continue }
            guard !existingUrls.contains(url) else { continue }

            existingNames.insert(name.lowercased())
            existingUrls.insert(url)

            let country = obj["country"] as? String ?? ""
            let tags    = obj["tags"]    as? String ?? ""
            result.append(RadioStation(
                id: nextId,
                name: name,
                streamUrl: url,
                description: country.isEmpty ? tags : country
            ))
            nextId += 1
        }
        return result
    }
}
